src/Psalm/Checker/Statements/Expression/Call/MethodCallChecker.php

```diff
@@ -542,4 +553,75 @@ public static function analyze(
             $context->vars_in_scope[$var_id] = $class_type;
         }
     }
+
+    /**
+     * If a magic getter or setter is used to access a property on a class that is not defined
+     * but a `@property` annotation for the property exists, throw an error. If no `@property`
+     * annotation exists, it's not an error because you're allowed to make magic getters and
+     * setters do crazy things.
+     *
+     * @param StatementsChecker $statements_checker
+     * @param \Psalm\Checker\ProjectChecker $project_checker
+     * @param PhpParser\Node\Expr\MethodCall $stmt
+     * @param string $fq_class_name
+     *
+     * @return bool
+     */
+    private static function checkMagicGetterOrSetterUndefinedProperty(
+        StatementsChecker $statements_checker,
+        \Psalm\Checker\ProjectChecker $project_checker,
+        PhpParser\Node\Expr\MethodCall $stmt,
+        string $fq_class_name
+    ) {
+        if (!is_string($stmt->name)) {
+            return true;
+        }
+
+        $method_name = strtolower($stmt->name);
+        if (!in_array($method_name, ['__get', '__set'], true)) {
+            return true;
+        }
+
+        $first_arg_value = $stmt->args[0]->value;
+        if (!$first_arg_value instanceof PhpParser\Node\Scalar\String_) {
+            return true;
+        }
+
+        $prop_name = $first_arg_value->value;
+        $property_id = $fq_class_name . '::$' . $prop_name;
+        $class_storage = $project_checker->classlike_storage_provider->get($fq_class_name);
+
+        switch ($method_name) {
+            case '__set':
+                if (!ClassLikeChecker::propertyExists($project_checker, $property_id)
```

Comment:
You can move this check above the switch since it's shared. Also I subtly changed the API so it's now `$project_checker->codebase->properties->propertyExists($property_id)`

Reply:
tricky tricky. will do.